lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,271 @@
+}
+
+// Creates a volume directory and volume object for NFS based volumes
+func (v *VolumeStore) VolumeCreate(op trace.Operation, ID string, store *url.URL, capacityKB uint64, info map[string][]byte) (*storage.Volume, error) {
+	target, err := v.Service.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+	defer v.Service.Unmount(target)
+
+	if _, err := target.Mkdir(v.volDirPath(ID), defaultPermissions); err != nil {
+		return nil, err
+	}
+
+	backing := NewNFSVolumeDevice(v.Target, v.volDirPath(ID))
+
+	vol, err := storage.NewVolume(v.SelfLink, ID, info, backing)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := writeMetadata(op, v.volMetadataDirPath(ID), info, target); err != nil {
+		return nil, err
+	}
+
+	op.Infof("nfs volume (%s) successfully created on volume store (%s)", ID, v.Name)
+	return vol, nil
+}
+
+// Removes a volume and all of its contents from the nfs store. We already know via the cache if it is in use.
+func (v *VolumeStore) VolumeDestroy(op trace.Operation, vol *storage.Volume) error {
+	target, err := v.Service.Mount(v.Target)
+	if err != nil {
+		return err
+	}
+	defer v.Service.Unmount(target)
+
+	op.Infof("Attempting to remove volume (%s) and its metadata from volume store (%s)", vol.ID, v.Name)
+
+	//remove volume directory and children
+	if err := target.RemoveAll(v.volDirPath(vol.ID)); err != nil {
+		op.Errorf("failed to remove volume (%s) on volume store (%s) due to error (%s)", vol.ID, v.Name, err)
+		return err
+	}
+
+	//XXX: what should we do if we lose the nfs server between volume and metadata deletion?
+
+	//remove volume metadata directory and children
+	if err := target.RemoveAll(v.volMetadataDirPath(vol.ID)); err != nil {
+		op.Errorf("failed to remove metadata for volume (%s) at path (%q) on volume store (%s)", vol.ID, v.volDirPath(vol.ID), v.Name)
+		//FIXME: Should we bail here? the volume is gone at this point...
+	}
+	op.Infof("Successfully removed volume (%s) from volumestore (%s)", vol.ID, v.Name)
+
+	return nil
+}
+
+func (v *VolumeStore) VolumesList(op trace.Operation) ([]*storage.Volume, error) {
+
+	target, err := v.Service.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+	defer v.Service.Unmount(target)
+
+	volFileInfo, err := target.ReadDir(v.volumesDir())
+	if err != nil {
+		return nil, err
+	}
+	var volumes []*storage.Volume
+	var fetchErrors []error
+
+	for _, fileInfo := range volFileInfo {
+
+		if fileInfo.Name() == "." || fileInfo.Name() == ".." {
+			continue
+		}
+
+		volMetadata, err := getMetadata(op, v.volMetadataDirPath(fileInfo.Name()), target)
+		if err != nil {
+			fetchErrors = append(fetchErrors, err)
+			continue
+		}
+
+		volDeviceBacking := NewNFSVolumeDevice(v.Target, v.volDirPath(fileInfo.Name()))
+
+		vol, err := storage.NewVolume(v.SelfLink, fileInfo.Name(), volMetadata, volDeviceBacking)
+		if err != nil {
+			return nil, err
+		}
+
+		volumes = append(volumes, vol)
+	}
+
+	if len(fetchErrors) > 0 {
```

Comment:
Really, just returning the last error is probably enough in this case.  We just want the caller to fail.  And if it's a systemic issue, you'll have to look at the logs anyway and provided you add the log line above, that should give the user enough chronology to figure out what happened in what order.